### PR TITLE
Z-150 Blackheart refactor

### PR DESCRIPTION
**Pull request by noroadsleft**
## Description

Modernizes the Z-150 Blackheart codebase and add QMK Configurator rendering support.

## Notes

tagging @blindassassin111 (PCB designer/maintainer)

### CODE REVIEW

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 1]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
 
 bool process_record_user(uint16_t keycode, keyrecord_t *record) {
-	return true;
+    return true;
 }
 
 void led_set_user(uint8_t usb_led) {
-  DDRB |= (1 << 0);
-  DDRE |= (1 << 6) | (1 << 7);
+    DDRB |= (1 << 0);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
    setPinOutput(B0);
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 2]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+    DDRE |= (1 << 6) | (1 << 7);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
    setPinOutput(E6);
    setPinOutput(E7);
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 3]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
 
-  if (usb_led & (1 << USB_LED_NUM_LOCK)) {
-    PORTE |= (1 << 7);
-  } else {
-    PORTE &= ~(1 << 7);
-  }
+    if (usb_led & (1 << USB_LED_NUM_LOCK)) {
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_NUM_LOCK)) {
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 4]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+        PORTE |= (1 << 7);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
        writePinHigh(E7);
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 5]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+    } else {
+        PORTE &= ~(1 << 7);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
        writePinLow(E7);
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 6]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+    }
 
-  if (usb_led & (1 << USB_LED_CAPS_LOCK)) {
-    PORTB |= (1 << 0);
-  } else {
-    PORTB &= ~(1 << 0);
-  }
+    if (usb_led & (1 << USB_LED_CAPS_LOCK)) {
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_CAPS_LOCK)) {
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 7]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+        PORTB |= (1 << 0);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
        writePinHigh(B0);
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 8]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+    } else {
+        PORTB &= ~(1 << 0);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
        writePinLow(B0);
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 9]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+    }
 
-  if (usb_led & (1 << USB_LED_SCROLL_LOCK)) {
-    PORTE |= (1 << 6);
-  } else {
-    PORTE &= ~(1 << 6);
-  }
+    if (usb_led & (1 << USB_LED_SCROLL_LOCK)) {
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_SCROLL_LOCK)) {
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 10]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+        PORTE |= (1 << 6);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
        writePinHigh(E6);
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 11]
```diff
@@ -42,28 +41,28 @@ void matrix_scan_user(void) {
...
+    } else {
+        PORTE &= ~(1 << 6);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
        writePinLow(E6);
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+void matrix_scan_user(void) {
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    return true;
+}
+
+void led_set_user(uint8_t usb_led) {
+    DDRB |= (1 << 0);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
    setPinOutput(B0);
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    DDRE |= (1 << 6) | (1 << 7);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
    setPinOutput(E6);
    setPinOutput(E7);
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+    if (usb_led & (1 << USB_LED_NUM_LOCK)) {
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_NUM_LOCK)) {
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        PORTE |= (1 << 7);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
        writePinHigh(E7);
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    } else {
+        PORTE &= ~(1 << 7);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
        writePinLow(E7);
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    if (usb_led & (1 << USB_LED_CAPS_LOCK)) {
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_CAPS_LOCK)) {
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+        PORTB |= (1 << 0);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
        writePinHigh(B0);
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    } else {
+        PORTB &= ~(1 << 0);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
        writePinLow(B0);
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    if (usb_led & (1 << USB_LED_SCROLL_LOCK)) {
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_SCROLL_LOCK)) {
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        PORTE |= (1 << 6);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
        writePinHigh(E6);
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    } else {
+        PORTE &= ~(1 << 6);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
        writePinLow(E6);
```

**Comment by fauxpark** [comment]
Actually, I wonder if the LED code should go in `led_set_kb()` instead...

**Comment by noroadsleft** [comment]
@fauxpark I was wondering exactly that myself. I noticed yesterday the blocks were identical, and I actually wanted to use the GPIO commands, but I don't actually understand them exactly. I compiled a hex that I `md5`'d, changed the default to use the GPIO commands, recompiled and `md5`'d again, *and* got a different md5sum and a different avr-size.

This does seem like behavior that should be defined at the keyboard level though.

**Comment by fauxpark** [comment]
It's just macros all the way down - try only changing one of them, it *should* compile to the same hex.

I suspect the difference comes from the fact that each GPIO command has to be done individually, whereas if you manipulate the registers directly you can OR the bits together. Evidently the compiler isn't smart enough to combine consecutive `DDRn` operations to produce the same binary, or perhaps that's by design, considering it would break the JTAG disable procedure (two writes to `MCUCR.JTD` within 4? cycles).

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+
+void matrix_scan_user(void) {
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    return true;
+}
+
+void led_set_user(uint8_t usb_led) {
+    setPinOutput(B0);
```
Comment (drashna) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]: The `setPinOutput`'s should be in the `matrix_init_user` function, not here.

Comment (noroadsleft) [reply keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]: Any insight to elevating these to keyboard level (`matrix_init_kb`, supposedly)?

Comment (drashna) [reply keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]: I'm personally not a fan of doing it, because it makes it harder for the user to override. 
The flip side, is that moving it there means that it is functional when using the QMK Configurator.

So.... I think it would be best to move to the keyboard level, yeah.

Comment (fauxpark) [reply keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]: All these overrideable functions should really be doing something similar to `process_record_*()`, returning a boolean which the higher-level functions use to determine whether or not to proceed with their own code. But changing all those signatures would touch just about every keyboard :(

**Comment by blindassassin111** [comment]
Are y'all mainly just updating the board to use the newer QMK preferred syntax? Got so many messages it is hard to follow now.

**Comment by noroadsleft** [comment]
@blindassassin111 That's the idea.

Sorry about all the messages. :frowning_face:

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c (hunk 23 above)
Comment (drashna) [reply keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]: Probably. Though, that's a massive change...  But it may be worth doing, long term.

**Comment by drashna** [comment]
@blindassassin111 Sorry about that.  It just brought up a good issue with how things are handled, and discussion about improving this functionality, especially to make it more consistent with other functionality.

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c (hunk 23 above)
Comment (drashna) [reply keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]: But for now, yeah, I think the led code should be moved to the kb functions, instead.

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 24]
```diff
@@ -1,38 +1,37 @@
...
+    [1] = LAYOUT(
+        _______, _______,   RESET,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,   _______, _______, _______,
+        _______, _______,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,   _______, _______, _______, _______,
+        _______, _______,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,                     _______, _______, _______, _______,
+        _______, _______,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,            _______, _______, _______, _______,
+        _______, _______,   _______, _______,                            _______,                                                       _______,                _______,          _______
+    ),
 };
 
 void matrix_init_user(void) {
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
void matrix_init_user(void) {
    setPinOutput(B0);
    setPinOutput(E6);
    setPinOutput(E7);
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 25]
```diff
@@ -42,28 +41,29 @@ void matrix_scan_user(void) {
...
 
 bool process_record_user(uint16_t keycode, keyrecord_t *record) {
-	return true;
+    return true;
 }
 
 void led_set_user(uint8_t usb_led) {
-  DDRB |= (1 << 0);
-  DDRE |= (1 << 6) | (1 << 7);
+    setPinOutput(B0);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 26]
```diff
@@ -42,28 +41,29 @@ void matrix_scan_user(void) {
...
+    setPinOutput(E6);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
```

File: keyboards/z150_blackheart/keymaps/default/keymap.c [hunk 27]
```diff
@@ -42,28 +41,29 @@ void matrix_scan_user(void) {
...
+    setPinOutput(E7);
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default/keymap.c]:
```suggestion
```

File: keyboards/z150_blackheart/keymaps/default_tkl/keymap.c [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+    [1] = LAYOUT_z150_tkl(
+        _______, _______,   RESET,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,     _______, _______, _______,
+        _______, _______,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,     _______, _______, _______,
+        _______, _______,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,              _______, _______, _______,
+        _______, _______,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,     _______, _______, _______,
+        _______, _______,   _______, _______, _______,                            _______,                            _______, _______, _______, _______,     _______, _______, _______
+    ),
+};
+
+void matrix_init_user(void) {
```
Comment (fauxpark) [inline keyboards/z150_blackheart/keymaps/default_tkl/keymap.c]:
```suggestion
void matrix_init_user(void) {
    setPinOutput(B0);
    setPinOutput(E6);
    setPinOutput(E7);
```

**Comment by blindassassin111** [comment]
@noroadsleft @drashna Y'all are good was just confusing when github sent 31 emails for everything and I had no idea why, lol. Happy to see that QMK has moved towards using much easier to code functions though!

**Comment by fauxpark** [comment]
The email spam (from me, anyway) is an unfortunate side effect of the limitations of GitHub's review suggestion system. You can only do make suggestions on a single line (as opposed to a range), and you can't batch them together to make them part of the same suggestion 😢

**Comment by noroadsleft** [comment]
I believe that I have properly and successfully done the thing.

**Comment by blindassassin111** [comment]
> The email spam (from me, anyway) is an unfortunate side effect of the limitations of GitHub's review suggestion system. You can only do make suggestions on a single line (as opposed to a range), and you can't batch them together to make them part of the same suggestion 😢
> 
> EDIT: This is actually wrong, and I've been pressing the wrong button this whole time. Sorry.

You are good, just confusing when I had no idea it was going to happen. Can't be mad at all, not sure if I should be happy that my keymap sparked the conversation though...not sure if that says something good or bad about it lol.

**Comment by drashna** [comment]
Looks good, thanks!
